### PR TITLE
Modified AbstractPlatform::getLocateExpression() and ::getSubstringExpression() signatures

### PR DESCRIPTION
**Pull request by morozov**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | yes

This breaking change is needed to enforce the type of `$start` natively. Additionally, all function arguments that accept an SQL expression will be type-hinted as `string` or `?string`, regardless of which SQL type they represent. Function arguments have been rename closer to their SQL analogs.

### CODE REVIEW

File: lib/Doctrine/DBAL/Platforms/PostgreSqlPlatform.php [hunk 1]
```diff
@@ -75,18 +75,6 @@ public function setUseBooleanTrueFalseStrings($flag)
         $this->useBooleanTrueFalseStrings = (bool) $flag;
     }
 
-    /**
-     * {@inheritDoc}
-     */
-    public function getSubstringExpression($value, $from, $length = null)
```
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/PostgreSqlPlatform.php]: This is the same as the default implementation.

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php [hunk 2]
```diff
@@ -96,28 +96,26 @@ public function getTrimExpression(string $str, int $mode = TrimMode::UNSPECIFIED
 
     /**
      * {@inheritDoc}
-     *
-     * SQLite only supports the 2 parameter variant of this function
```
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: This is obsolete.

File: tests/Doctrine/Tests/DBAL/Platforms/DB2PlatformTest.php [hunk 3]
```diff
@@ -332,7 +332,6 @@ public function testGeneratesSQLSnippets()
         self::assertEquals("'1987/05/02' - 10 YEAR", $this->platform->getDateSubYearsExpression("'1987/05/02'", 10));
         self::assertEquals(' WITH RR USE AND KEEP UPDATE LOCKS', $this->platform->getForUpdateSQL());
         self::assertEquals('LOCATE(substring_column, string_column)', $this->platform->getLocateExpression('string_column', 'substring_column'));
-        self::assertEquals('LOCATE(substring_column, string_column)', $this->platform->getLocateExpression('string_column', 'substring_column'));
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Platforms/DB2PlatformTest.php]: This is a duplicate of the above.

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php [hunk 4]
```diff
@@ -850,17 +850,15 @@ public function getLowerExpression($str)
...
     /**
-     * Returns the SQL snippet to get the position of the first occurrence of substring $substr in string $str.
+     * Returns the SQL snippet to get the position of the first occurrence of the substring in the string.
      *
-     * @param string    $str      Literal string.
-     * @param string    $substr   Literal string to find.
-     * @param int|false $startPos Position to start at, beginning of string by default.
-     *
-     * @return string
+     * @param string      $string    String to locate the substring in, SQL expression.
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: This comment is unclear to me - can you rephrase it, please?

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php [hunk 5]
```diff
@@ -850,17 +850,15 @@ public function getLowerExpression($str)
...
+     * @param string      $substring Substring to locate, SQL expression.
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: I think a better naming could be "needle"

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php [hunk 6]
```diff
@@ -850,17 +850,15 @@ public function getLowerExpression($str)
...
+     * @param string|null $start     Position to start at, SQL expression. Defaults to the beginning of the string.
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: Alao confusing - why is this a `string`? Could we maybe add example usages?

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php [hunk 7]
```diff
@@ -876,25 +874,22 @@ public function getNowExpression()
...
      *
-     * SQLite only supports the 2 parameter variant of this function.
-     *
-     * @param string   $value  An sql string literal or column name/alias.
-     * @param int      $from   Where to start the substring portion.
-     * @param int|null $length The substring portion length.
-     *
-     * @return string
+     * @param string      $string String to get a substring in, SQL expression.
+     * @param string      $start  Position to start at, SQL expression.
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: Similar to the above, `start` being a `string` is confusing to me - we'll need usage examples, IMO

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php (hunk 2 above)
Comment (Ocramius) [reply lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: Do we know since when? We should probably require SQLite to be at least newer than X

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php (hunk 4 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: This is the best I could come up with. Otherwise, it was "The SQL expression which represents the string in which the substring should be located". And this "the SQL expression" will be in a lot of places. Is that better?

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php (hunk 5 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: I'd keep it closer to the SQL naming, not to PHP. Otherwise, it'd have to be `$haystack` and `$needle`.

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php (hunk 6 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: Because it's an SQL expression, not a value (see the test). It could a reference to another column in the query or a function call or whatever else. There will be a lot of changes like this in the API. Not sure if we want to put an example in every occurrence. What would you suggest?

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php (hunk 4 above)
Comment (Ocramius) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: Aha, the wording "sql expression" put after the comma did confuse me

**Comment by morozov** [comment]
> Do we know since when? We should probably require SQLite to be at least newer than X

No, but [at least as of 2008](https://web.archive.org/web/20081219142839/https://www.sqlite.org/lang_corefunc.html).

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php (hunk 4 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: So what do you recommend?

Comment (Ocramius) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: "SQL expression producing the haystack" or something like that?

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php (hunk 7 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: There are more than 40 methods in this class which accept SQL expressions. We do need some explanation but it doesn't have to be a phpDoc example for every one of them. Maybe it could be some footnote or a link to the online documentation?

I'd like to leave the doc block in a good enough shape and proceed with the changes. I hope, someone could join later and improve the documentation as well.

**Comment by morozov** [comment]
@Ocramius do you still have any concerns or we can proceed?

**Review by Ocramius — APPROVED** [review]
Let's proceed 👍
